libs/wire-api-federation/src/Wire/API/Federation/Client.hs

```diff
@@ -58,26 +59,37 @@ instance KnownComponent 'Proto.Galley where
 
 -- | expectedStatuses is ignored as we don't get any status from the federator,
 -- all responses have '200 OK' as their status.
-instance (Monad m, MonadError FederationClientError m, MonadIO m, KnownComponent component) => RunClient (FederatorClient component m) where
+instance (Monad m, MonadIO m, MonadError FederationClientFailure m, KnownComponent component) => RunClient (FederatorClient component m) where
   runRequestAcceptStatus _expectedStatuses req = do
     env <- ask
+    let path = LBS.toStrict . toLazyByteString $ requestPath req
+        domain = targetDomain env
+        mkFailure = FederationClientFailure domain path
+        failure :: MonadError FederationClientFailure n => FederationClientError -> n x
+        failure = throwError . mkFailure
+        rpcFailure = failure . FederationClientRPCError
+        readBody = \case
+          RequestBodyLBS lbs -> pure $ LBS.toStrict lbs
+          RequestBodyBS bs -> pure bs
+          RequestBodySource _ -> failure FederationClientStreamingUnsupported
+    put (Just path)
```

Comment:
Hm, does this make an actual put request? Shouldn't this only happen a few lines below in `callRemote` ?

Reply:
Good question, I should have explained what this is in a comment. It is not a PUT request, but simply saving the path in the state of the monad stack. The function `put` is [this one](http://hackage.haskell.org/package/mtl-2.2.2/docs/Control-Monad-State-Class.html).

The idea here is that Servant is going to call this function for each request, and call `throwClientError` when some error occurs while translating a response into a Servant response. Since the path we want to add to the error structure is contained in the request, we have no direct way to access it from our implementation of `throwClientError` below.

My solution was to change the monad to keep a `Maybe ByteString` in its state, set it when a request is received, and fetch it when we need to produce an error. As far as I understand, Servant guarantees that `throwClientError` is only called after `runRequestAcceptStatus`, so this should work fine. If for any reason this guarantee does not hold, we get an empty path, but nothing catastrophic.

Reply:
Ah, that makes more sense. Too many functions have the same name and without a qualified import it's not directly obvious. Thanks for the explanation and the comments!